dkg/src/dkg_manager/mod.rs

```diff
@@ -40,29 +95,199 @@ impl<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> DKGManager<DKG, P> {
+        if let InnerState::Finished {
+            start_time_us,
+            pull_confirmed,
+            ..
+        } = &mut self.state
+        {
+            if !*pull_confirmed {
+                // TODO(zjma): metric DKG_AGG_NODE_PROPOSED
+            }
+            *pull_confirmed = true;
+        }
+        Ok(())
+    }
+
+    /// Calculate DKG config. Deal a transcript. Start broadcasting the transcript.
+    /// Called when a DKG start event is received, or when the node is restarting.
+    ///
+    /// NOTE: the dealt DKG transcript does not have to be persisted:
+    /// it is ok for a validator to equivocate on its DKG transcript, as long as the transcript is valid.
+    async fn setup_deal_broadcast(
+        &mut self,
+        start_time_us: u64,
+        target_validator_set: &ValidatorSet,
+    ) -> Result<()> {
+        self.state = match &self.state {
+            InnerState::NotStarted => {
+                let public_params = DKG::new_public_params(
+                    self.epoch_state.as_ref(),
+                    self.my_addr,
+                    target_validator_set,
+                );
+                let mut rng = thread_rng();
+                let trx = DKG::generate_transcript(
+                    &mut rng,
+                    self.private_params_provider.dkg_private_params(),
+                    &public_params,
+                );
+
+                let dkg_node = DKGNode::new(
+                    self.epoch_state.epoch,
+                    self.my_addr,
+                    DKG::serialize_transcript(&trx),
+                );
+
+                // TODO(zjma): DKG_NODE_READY metric
+
+                let abort_handle = self.agg_trx_producer.start_produce(
+                    self.epoch_state.clone(),
+                    public_params.clone(),
+                    self.agg_trx_tx.clone(),
+                );
+
+                // Switch to the next stage.
+                InnerState::InProgress {
+                    start_time_us,
+                    public_params,
+                    my_node: dkg_node,
+                    abort_handle,
+                }
+            },
+            _ => unreachable!(), // `setup_deal_broadcast` is called only when DKG state is `NotStarted`.
+        };
+
+        Ok(())
+    }
+
+    /// On a locally aggregated transcript, put it into the validator txn pool and update inner states.
+    async fn process_aggregated_transcript(&mut self, agg_trx: DKG::Transcript) -> Result<()> {
+        self.state = match std::mem::take(&mut self.state) {
+            InnerState::InProgress {
+                start_time_us,
+                my_node,
+                ..
+            } => {
+                // TODO(zjma): metric DKG_AGG_NODE_READY
+                let txn = ValidatorTransaction::DKGResult(DKGNode {
+                    metadata: DKGTranscriptMetadata {
+                        epoch: self.epoch_state.epoch,
+                        author: self.my_addr,
```

Comment:
it looks different node will produce different DKG transaction? we don't deduplicate from them?

Reply:
that's the cost of we saving a DB. It's hard to duplicate now.